### PR TITLE
(PUP-5538) Improve SID::Principal.lookup_account_sid validation

### PR DESCRIPTION
**Pull request by glennsarti**
Previously the lookup_account_sid function would allow any
type of argument when calling lookup_account_sid, however this would
cause inconsistent error messages and possible memory errors when using
zero byte arrays or nils.  This commit adds validation checking to
sid_bytes before allocating memory pointers and adds tests for these
scenarios.

### CODE REVIEW

File: lib/puppet/util/windows/principal.rb [hunk 1]
```diff
@@ -87,6 +87,10 @@ def self.lookup_account_name(system_name = nil, account_name)
 
     def self.lookup_account_sid(system_name = nil, sid_bytes)
```
Comment (Iristyle) [inline lib/puppet/util/windows/principal.rb]: Should `system_name` get some validation also?

Comment (glennsarti) [inline lib/puppet/util/windows/principal.rb]: All of the bad cases will be trapped by the first call which converts it to a wide string. Even an empty string will still be two bytes long due to the encoding.  The same issue of zero length pointers will not occur.

Comment (Iristyle) [inline lib/puppet/util/windows/principal.rb]: :+1:

**Comment by Iristyle** [comment]
I think we should tack on a quick test for the `:SidTypeDomain`resolution as well in another commit - i.e. call `lookup_account_name` using an empty string for `account_name` / and lookup_account_sid with the byte equivalent of `S-1-5-32

**Comment by Iristyle** [comment]
Minor - might want to reword commit subject to something like `(PUP-5538) Improve lookup_account_sid validation` since we're handling an edge case rather than fixing something that's completely broken.

**Comment by glennsarti** [comment]
Added BUILTIN Domain test cases

File: spec/integration/util/windows/principal_spec.rb [hunk 2]
```diff
@@ -111,5 +142,12 @@
         principal.lookup_account_sid([1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0])
       }.to raise_error(Puppet::Util::Windows::Error, /Failed to call LookupAccountSidW:  No mapping between account names and security IDs was done/)
     end
+    
+    it "should return a domain principal for BUILTIN SID S-1-5-32" do
+      principal = Puppet::Util::Windows::SID::Principal.lookup_account_sid([1, 1, 0, 0, 0, 0, 0, 5, 32, 0, 0, 0])
+      expect(principal.account_type).to eq(:SidTypeDomain)
+      expect(principal.sid).to eq('S-1-5-32')
+      expect(principal.domain).to eq('BUILTIN')
+    end
```
Comment (Iristyle) [inline spec/integration/util/windows/principal_spec.rb]: Ah yeah, but what about looking up `BUILTIN` by name?  we should see that `account_domain` is hosed as `BUILTIN\BUILTIN` - which is the part I was interested in fixing now (one extra commit is fine to address that)

Comment (glennsarti) [inline spec/integration/util/windows/principal_spec.rb]: @Iristyle Extra commit added for the account_domain generation and tests

**Comment by Iristyle** [comment]
Follow-up commit at https://github.com/puppetlabs/puppet/commit/d376814db9c8110517272c778041a3ce7004332a
